examples/codes/external_code.ex.py

```diff
@@ -0,0 +1,428 @@
+#   external code
+# * A `RunMode` class to specify the possibly running modes
+# * A `Solver` that orchistrates the running of the code
+# * Some task for the `Solver` to run typically there are three:
+#     * A `Setup` task which writes the input file for the code
+#     * A `Run` task which runs the code
+#     * A `Teardown` task which reads the output file of the code
+#
+# The `MappedParameterFrame` here gets the defaults and sets the mappings.
+# Notice that "param2" is sent to the code but "param1" is not.
+
+# %%
+@dataclass
+class ECParameterFrame(MappedParameterFrame):
+    """External Code ParameterFrame"""
+
+    header: Parameter[bool]
+    line_number: Parameter[bool]
+    param1: Parameter[float]
+    param2: Parameter[float]
+
+    _defaults = (ECOpts(), ECInputs())
+
+    _mappings = {
+        "header": ParameterMapping("add_header", send=True, recv=False),
+        "line_number": ParameterMapping("number", send=True, recv=False),
+        "param1": ParameterMapping("param1", send=False, recv=True, unit="MW"),
+        "param2": ParameterMapping("param2", send=True, recv=True, unit="GW"),
+    }
+
+    @property
+    def mappings(self) -> Dict:
+        """Code Mappings"""
+        return self._mappings
+
+    @classmethod
+    def from_defaults(cls) -> MappedParameterFrame:
+        """Setup from defaults"""
+        dd = {}
+        for _def in cls._defaults:
+            dd = {**dd, **asdict(_def)}
+        return super().from_defaults(dd)
+
+
+# %%
+class RunMode(BaseRunMode):
+    """
+    RunModes for external code
+    """
+
+    RUN = auto()
+    READ = auto()
+    MOCK = auto()
+
+
+# %% [markdown]
+# The `Setup` class pulls over the inputs from bluemira as described by the
+# mapping and creates the input file. The output of the run method returns
+# the commandline options list.
+
+# %%
+class Setup(CodesSetup):
+    """Setup task"""
+
+    params: ECParameterFrame
+
+    def __init__(self, params: ParameterFrame, problem_settings: Dict, infile: str):
+        super().__init__(params, PRG_NAME)
+        self.problem_settings = problem_settings
+        self.infile = infile
+
+    def update_inputs(self) -> Dict:
+        """Update inputs from bluemira"""
+        self.inputs = ECInputs()
+        self.options = ECOpts()
+        inp = self._get_new_inputs()
+        for thing in (self.problem_settings.items(), inp.items()):
+            for k, v in thing:
+                if k in self.options.__annotations__:
+                    if v:
+                        setattr(self.options, k, v)
+                    else:
+                        continue
+                elif v is not None:
+                    getattr(self.inputs, k)
+                    setattr(self.inputs, k, v)
+
+        return {k: v for k, v in asdict(self.inputs).items() if v}
+
+    def run(self) -> List:
+        """Run mode"""
+        inp = self.update_inputs()
+        with open(self.infile, "w") as _if:
```

Comment:
```suggestion
        with open(self.infile, "w") as input_file:
```

I think this is a bit more descriptive, and then we don't need the leading `_` on the `if` - I was initially a bit confused why we needed a private variable here, until I realised it was to allow calling the variable `if`